src/graph/graph.c

```diff
@@ -680,10 +674,288 @@ void Graph_DeleteNode(Graph *g, Node *n) {
+            }
+        }
+
+        // Free and remove edges from datablock.
+        DataBlock_DeleteItem(g->edges, ENTITY_GET_ID(e));
+    }
+
+    // Delete entries.
+    for(int i = 0; i < array_len(deletions); i++) {
+        deletion = deletions[i];
+        assert(GxB_Matrix_Delete(deletion.M, deletion.row, deletion.col) == GrB_SUCCESS);
+    }
+
+    int relationCount = Graph_RelationTypeCount(g);
+    uint deletion_count = array_len(deletions);
+    for(uint i = 0; i < deletion_count; i++) {
+        deletion = deletions[i];
+        GrB_Index src = deletion.col;
+        GrB_Index dest = deletion.row;
+
+        // See if source is connected to destination with additional edges.
+        bool connected = false;
+        for(int i = 0; i < relationCount; i++) {
+            R = Graph_GetRelationMatrix(g, i);
+            info = GrB_Matrix_extractElement_BOOL(&connected, R, dest, src);
+            if(info == GrB_SUCCESS) break;
+        }
+
+        /* There are no additional edges connecting source to destination
+         * Remove edge from THE adjacency matrix. 
+         * It is OK to remove entries from the adjacency matrix, as we're 
+         * not trying to extract entries from it. */
+        if(!connected) {
+            M = Graph_GetAdjacencyMatrix(g);
+            assert(GxB_Matrix_Delete(M, dest, src) == GrB_SUCCESS);
+
+            M = _Graph_Get_Transposed_AdjacencyMatrix(g);
+            assert(GxB_Matrix_Delete(M, src, dest) == GrB_SUCCESS);
+        }
+    }
+}
+
+/* Removes both nodes and edges from graph. */
+void Graph_BulkDelete(Graph *g, Node *nodes, size_t node_count, Edge *edges, size_t edge_count, size_t *edge_deleted) {
+    assert(g);
+    
+    *edge_deleted = 0;
+
+    if(node_count) _BulkDeleteNodes(g, nodes, node_count, edge_deleted);
+
+    if(edge_count) {
+        // Filter out explicit edges which were removed by _BulkDeleteNodes.
+        if(node_count) {
+            for(int i = 0; i < edge_count; i++) {
+                Edge *e = edges+i;
+                NodeID src = Edge_GetSrcNodeID(e);
+                NodeID dest = Edge_GetDestNodeID(e);
+                
+                if(!DataBlock_GetItem(g->nodes, src) || !DataBlock_GetItem(g->nodes, dest)) {
+                    /* Edge already removed due to node removal.
+                    * Replace current edge with last edge. */
+                    edges[i] = edges[edge_count-1];
+
+                    // Update indices.
+                    i--;
+                    edge_count--;
+                }
+            }
+        }
+
+        // Removing duplicates
+        #define is_edge_lt(a, b) (ENTITY_GET_ID((a)) < ENTITY_GET_ID((b)))
+        QSORT(Edge, edges, edge_count, is_edge_lt);
+
+        size_t uniqueIdx = 0;
+        for(int i = 0; i < edge_count; i++) {
+            // As long as current is the same as follows.
+            while(i < edge_count && ENTITY_GET_ID(edges+i) == ENTITY_GET_ID(edges+i+1)) i++;
```

Comment:
Use:
`while(i < edge_count - 1 && ENTITY_GET_ID(edges+i) == ENTITY_GET_ID(edges+i+1)) i++;`
As edges+i+1 will be out of range on the last check.

Reply:
👍